File: ts/src/cmd/diff.ts
import type { RadStore } from '../store';

export function runDiff(store: RadStore): string {
  const operations = store.loadOplog();

  // Get visible writes
  const visibleWrites = operations.filter(o => o.status === 'visible' && o.type === 'write');

  if (visibleWrites.length === 0) {
    return 'no pending changes\n';
  }

  // Build a map of accepted state (file -> latest accepted content)
  const acceptedState = new Map<string, string>();
  const opsByTime = [...operations].sort((a, b) => a.timestamp - b.timestamp);

  for (const op of opsByTime) {
    if (op.status === 'accepted' && op.type === 'write') {
      // Extract file path from regionId (format: "file:start-end")
      const colonPos = op.regionId.indexOf(':');
      if (colonPos !== -1) {
        const filePath = op.regionId.substring(0, colonPos);
        acceptedState.set(filePath, op.content);
      }
    }
  }

  // Build output
  let output = '';
  for (const op of visibleWrites) {
    // Extract file path from regionId
    const colonPos = op.regionId.indexOf(':');
    const filePath = colonPos !== -1 ? op.regionId.substring(0, colonPos) : op.regionId;

    // Check if this is a new file
    const isNewFile = !acceptedState.has(filePath);

    if (isNewFile) {
      output += '--- (new file)\n';
      output += `+++ visible by ${op.participantId} (${op.id})\n`;
      output += `+ ${op.regionId}  "${op.content}"\n`;
    } else {
      const acceptedContent = acceptedState.get(filePath)!;
      output += `--- accepted (${filePath})\n`;
      output += `+++ visible by ${op.participantId} (${op.id})\n`;
      output += `-${acceptedContent}\n`;
      output += `+${op.content}\n`;
    }
    output += '\n';
  }

  return output;
}

File: rust/cli/src/cmd/diff.rs
use crate::store::RadStore;
use crate::types::{OpStatus, OpType};
use std::collections::HashMap;

pub fn run_diff(store: &RadStore) -> Result<String, String> {
    let oplog = store.load_oplog()?;
    let ops = oplog.get_all_operations();

    // Get visible writes
    let visible_writes: Vec<_> = ops.iter()
        .filter(|o| o.status == OpStatus::Visible && o.op_type == OpType::Write)
        .collect();

    if visible_writes.is_empty() {
        return Ok("no pending changes\n".to_string());
    }

    // Build a map of accepted state (file -> latest accepted content)
    let mut accepted_state: HashMap<String, String> = HashMap::new();
    let mut ops_by_time = ops.clone();
    ops_by_time.sort_by_key(|o| o.timestamp);

    for op in &ops_by_time {
        if op.status == OpStatus::Accepted && op.op_type == OpType::Write {
            // Extract file path from region_id (format: "file:start-end")
            if let Some(colon_pos) = op.region_id.find(':') {
                let file_path = op.region_id[..colon_pos].to_string();
                accepted_state.insert(file_path, op.content.clone());
            }
        }
    }

    // Build output
    let mut output = String::new();
    for op in visible_writes {
        // Extract file path from region_id
        let file_path = if let Some(colon_pos) = op.region_id.find(':') {
            &op.region_id[..colon_pos]
        } else {
            &op.region_id
        };

        // Check if this is a new file
        let is_new_file = !accepted_state.contains_key(file_path);

        if is_new_file {
            output.push_str("--- (new file)\n");
            output.push_str(&format!("+++ visible by {} ({})\n", op.participant_id, op.id));
            output.push_str(&format!("+ {}  \"{}\"\n", op.region_id, op.content));
        } else {
            let accepted_content = accepted_state.get(file_path).unwrap();
            output.push_str(&format!("--- accepted ({})\n", file_path));
            output.push_str(&format!("+++ visible by {} ({})\n", op.participant_id, op.id));
            output.push_str(&format!("-{}\n", accepted_content));
            output.push_str(&format!("+{}\n", op.content));
        }
        output.push_str("\n");
    }

    Ok(output)
}
